### DISCUSSION TITLE
Restore pool configuration to prior to incorrectly added device.

### DISCUSSION THREAD

**Discussion by shousper** [openzfs/zfs#14389] [Q&A]
Hello,

I've fallen victim to [this mistake](https://github.com/openzfs/zfs/discussions/11599), and before researching more carefully how to go about rectifying it, I fell victim to [this mistake](https://github.com/openzfs/zfs/discussions/11665) and now have an unimportable pool with an unavailable device. Worse still, I had already erased the partition table in preparation to move data around to re-create the pool before realising my first mistake.

Is it at all possible to revert the configuration to it's state prior to the device being incorrectly added? Nothing was written to the new device, it was never used, so I'm hoping there might be a way to rollback the configuration or manually hack it by hand?

This is the current state of the pool:

```
$ sudo zpool import -F
   pool: tank
     id: 11002558381027971913
  state: UNAVAIL
 status: One or more devices are faulted.
 action: The pool cannot be imported due to damaged devices or data.
 config:

        tank                                          UNAVAIL  insufficient replicas
          raidz1-0                                    ONLINE
            ata-WDC_WD40EFPX-68C6CN0_WD-WX12A82EDEVN  ONLINE
            ata-WDC_WD40EFPX-68C6CN0_WD-WX32A826DLN2  ONLINE
            ata-WDC_WD40EFPX-68C6CN0_WD-WX52D62A2D0V  ONLINE
          ata-WDC_WD40EFPX-68C6CN0_WD-WX32A82FTTCJ    FAULTED  corrupted data
```

Any attempt to import it with any combination of flags yields the same:

```
cannot import 'tank': one or more devices is currently unavailable
```

From the 2 discussions I referenced above, I could only try changing `zfs_max_missing_tvds` with `echo 1 > /sys/module/zfs/parameters/zfs_max_missing_tvds`, but for some reason I cannot write to it 🤔

I'm currently using a standard install of `zfs@0.8.3` installed via `apt` on `ubuntu@20.04` -- which I see is significantly behind the current release. If there is anything to be gained in upgrading to help resolve this, please let me know.

Thanks in advance 🙏🏻

**Comment by rincebrain** [comment]
That parameter doesn't exist on a ZFS version that ancient. Try a newer one.

**Reply by shousper** [reply]
Hi @rincebrain, thanks for replying 🙏🏻

I've installed `zfs@2.1.7` from source and tried to update the parameter -- which worked -- but the import still fails with the same error ☹️

```
$ ls -la /sys/module/zfs/parameters/zfs_max_missing_tvds
-rw-r--r-- 1 root root 4096 Jan 14 18:35 /sys/module/zfs/parameters/zfs_max_missing_tvds
$ echo 1 > /sys/module/zfs/parameters/zfs_max_missing_tvds
$ cat /sys/module/zfs/parameters/zfs_max_missing_tvds
1
$ sudo zpool import -F -f -m tank
cannot import 'tank': one or more devices is currently unavailable
$ zfs version
zfs-2.1.7-1
zfs-kmod-2.1.7-1
$
```

Is there any way to see and/or alter the underlying/internal configuration or state of the pool? Perhaps a development tool used for testing?

EDIT: I forgot `-o readonly=on` -- this worked to get my data back online and available.

```
$ zpool import -F -f -m -o readonly=on tank
$ zpool status
  pool: tank
 state: DEGRADED
status: One or more devices are faulted in response to persistent errors.
        Sufficient replicas exist for the pool to continue functioning in a
        degraded state.
action: Replace the faulted device, or use 'zpool clear' to mark the device
        repaired.
  scan: resilvered 684K in 00:00:01 with 0 errors on Thu Jan 12 22:48:15 2023
config:

        NAME                                          STATE     READ WRITE CKSUM
        tank                                          DEGRADED     0     0     0
          raidz1-0                                    ONLINE       0     0     0
            ata-WDC_WD40EFPX-68C6CN0_WD-WX12A82EDEVN  ONLINE       0     0     0
            ata-WDC_WD40EFPX-68C6CN0_WD-WX32A826DLN2  ONLINE       0     0     0
            ata-WDC_WD40EFPX-68C6CN0_WD-WX52D62A2D0V  ONLINE       0     0     0
          ata-WDC_WD40EFPX-68C6CN0_WD-WX32A82FTTCJ    FAULTED      0     0     0  external device fault

errors: No known data errors
$ zfs list
NAME            USED  AVAIL     REFER  MOUNTPOINT
tank           2.25T  8.51T      128K  /tank
$
```

Thank you so much 🙏🏻

I'd still really prefer not to have to re-create the pool again if avoidable. Even if only to understand ZFS better and/or identify a better solution than moving around large volumes of data (IMO) unnecessarily. Maybe it is necessary and I can learn why.

So my question above still stands: Is there a way to manually hack the configuration or state of the pool to reverse the incorrect addition of the "faulted" device?

**Reply by rincebrain** [reply]
It's not avoidable at this point, really.

Even if you could hack the configuration so the vdev list no longer mentioned the absent disk, there's lots of spots on disk that mention the vdev that no longer is there.

It will, in general, try to warn you not to do that unless you use -f, but if you use -f, nothing's going to stop you.

**Reply by shousper** [reply]
> Even if you could hack the configuration so the vdev list no longer mentioned the absent disk, there's lots of spots on disk that mention the vdev that no longer is there.

Sounds fun!

How many places on disk? Could updates potentially be scripted? How would one go about locating them all?

Where is the config specifying the vdev?

Sorry, that's a lot more questions 😅 I do really appreciate you taking the time though 🖖

**Reply by rincebrain** [reply]
Every block pointer on disk contains a reference to which vdev the data in it is from, and in turn, every metadata block has its checksum stored in the parent metadata block, and so on.

So you'd get to rewrite rather a lot of data to try and fudge this on-disk, and that's just off the top of my head.

**Reply by shousper** [reply]
Okay, could be a bit of a challenge, haha.

As I said, no data has been written to the accidentally added device -- at least not outside of `zpool` interactions.

When a device is added to the pool, are all the blocks and their metadata rewritten by the add (or attach?) subcommand to include the new vdev? Or is the vdev only added to blocks with data on the new device?

If it is rewritten when the command runs to add the new device, I could maybe reverse engineer that process from the source. Depending how complex it is.

If it's not rewritten, perhaps there are a lot less blocks, but I'd still need help identifying them..

Honestly, I may need help identifying a few things, like where in the source the block and metadata format is specified, and where the CLI command source links through to the code that actually writes to disk. I'm not familiar at all with how kernel modules work (or the current state of C development 😅), though I do have a bunch of time up my sleeve.

That being said, perhaps you're right and it's not worth the time versus recreating the pool. Especially with only 3 people potentially running into this scenario. I feel I've already taken too much of your time.

**Reply by Fatimah1986** [reply]
Yes I want to

**Comment by IvanVolosyuk** [comment]
You may have to have a look at ZFS on disk format. This is the first link
in Google on this topic: http://www.giis.co.in/Zfs_ondiskformat.pdf
You may also want to put sparse images of the disks of the broken pool on
ZFS and snapshot them first before doing experiments on the data. You may
need a lot of extra space for that. Have a look at zdb command if you
didn't do that yet.

**Reply by shousper** [reply]
Hi @IvanVolosyuk, thanks for the direction 🙏🏻

I may be wrong, but it appears `zdb` unfortunately only works with imported/importable pools? 🤷🏻 I tried a few options (`-C`, `-d`, `-S` from the manpage examples) and they all responded with: "zdb: can't open 'tank': No such file or directory"

The on disk format document looks quite helpful though, thanks!

**Comment by rincebrain** [comment]
-e is for exported pools.

**Reply by shousper** [reply]
Thanks, I'm currently waiting for `zpool import -F -X` to see if it can tell me anything useful.

**Comment by rincebrain** [comment]
You're aware zpool import -F -X is very destructive, right?

**Reply by shousper** [reply]
Yup 🙂 I've already pulled my data onto another disk so I could experiment with this problem. Looks like it's happily found a previous state it could import -- I'm not sure how to tell when it's from, but I can see it's from before I upgraded the features on the pool.

```
$ zpool status
  pool: tank
 state: ONLINE
status: Some supported and requested features are not enabled on the pool.
	The pool can still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
	the pool may no longer be accessible by software that does not support
	the features. See zpool-features(7) for details.
  scan: resilvered 684K in 00:00:01 with 0 errors on Thu Jan 12 22:48:15 2023
config:

	NAME                                          STATE     READ WRITE CKSUM
	tank                                          ONLINE       0     0     0
	  raidz1-0                                    ONLINE       0     0     0
	    ata-WDC_WD40EFPX-68C6CN0_WD-WX12A82EDEVN  ONLINE       0     0     0
	    ata-WDC_WD40EFPX-68C6CN0_WD-WX32A826DLN2  ONLINE       0     0     0
	    ata-WDC_WD40EFPX-68C6CN0_WD-WX52D62A2D0V  ONLINE       0     0     0

errors: No known data errors
```

I'm just running a checksum comparison of all the data (`rsync -avcP --dry-run /mnt/recovery /mnt/tank`) to see if it's rolled back far enough to lose more data, or if it's rolled back just far enough to correct the pool configuration by dropping the dud device.

**Comment by shousper** [comment]
It would appear `zpool import -F -X` has successfully rolled back the configuration without destroying any data. I was now able to successfully add the remaining device as a spare to the pool 👍🏻

I was surprised to learn one cannot add a device to an existing raidz configuration though? Not to worry.

Thanks for all the help 🙏🏻

**Reply by rincebrain** [reply]
Check out #12225 for that feature.
